esp32/mods/esp_espnow.c

```diff
@@ -0,0 +1,359 @@
+       if (n_args == 1) {
+        return send_cb_obj;
+    }
+
+    send_cb_obj = args[1];
+    return mp_const_none;
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_VAR_BETWEEN(espnow_on_send_obj, 1, 2, espnow_on_send);
+
+STATIC mp_obj_t espnow_on_recv(size_t n_args, const mp_obj_t *args) {
+//    mp_obj_framebuf_t *self = MP_OBJ_TO_PTR(args[0]);
+    if (n_args == 1) {
+        return recv_cb_obj;
+    }
+
+    recv_cb_obj = args[1];
+    return mp_const_none;
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_VAR_BETWEEN(espnow_on_recv_obj, 1, 2, espnow_on_recv);
+
+// pmk(primary_key)
+STATIC mp_obj_t espnow_pmk(mp_obj_t self_in, mp_obj_t key) {
+    uint8_t buf[ESP_NOW_KEY_LEN];
+    _get_bytes(key, ESP_NOW_KEY_LEN, buf);
+    ESPNOW_EXCEPTIONS(esp_now_set_pmk(buf));
+    return mp_const_none;
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_2(espnow_pmk_obj, espnow_pmk);
+
+// lmk(peer_mac, local_key)
+STATIC mp_obj_t espnow_lmk(mp_obj_t self_in, mp_obj_t addr, mp_obj_t key) {
+    mp_uint_t addr_len;
+    const uint8_t *addr_buf = (const uint8_t *)mp_obj_str_get_data(addr, &addr_len);
+    if (addr_len != ESP_NOW_ETH_ALEN) mp_raise_ValueError("addr invalid");
+    esp_now_peer_info_t peer;
+    ESPNOW_EXCEPTIONS(esp_now_get_peer(addr_buf, &peer));
+
+    // set peer lmk
+    bool encrypt = (key != mp_const_none);
+    bool re_add = (peer.encrypt != encrypt);
+    if (encrypt) _get_bytes(key, ESP_NOW_KEY_LEN, peer.lmk);
+    if (re_add) {
+        // workaround for calling esp_now_mod_peer() to 
+        // change encryption status will crash the system
+        peer.encrypt = encrypt;
+        ESPNOW_EXCEPTIONS(esp_now_del_peer(addr_buf));
+        ESPNOW_EXCEPTIONS(esp_now_add_peer(&peer));
+    } else {
+        ESPNOW_EXCEPTIONS(esp_now_mod_peer(&peer));
+    }
+    return mp_const_none;
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_3(espnow_lmk_obj, espnow_lmk);
+
+// add_peer(peer_mac, [local_key])
+STATIC mp_obj_t espnow_add_peer(size_t n_args, const mp_obj_t *args) {
+    esp_now_peer_info_t peer = {0};
+    _get_bytes(args[1], ESP_NOW_ETH_ALEN, peer.peer_addr);
+    if (n_args > 2) {
+        _get_bytes(args[2], ESP_NOW_KEY_LEN, peer.lmk);
+        peer.encrypt = 1;
+    }
+
+    ESPNOW_EXCEPTIONS(esp_now_add_peer(&peer));
+    return mp_const_none;
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_VAR_BETWEEN(espnow_add_peer_obj, 2, 3, espnow_add_peer);
+
+// del_peer(peer_mac)
+STATIC mp_obj_t espnow_del_peer(mp_obj_t self_in, mp_obj_t addr) {
+    mp_uint_t addr_len;
+    const uint8_t *addr_buf = (const uint8_t *)mp_obj_str_get_data(addr, &addr_len);
+    if (addr_len != ESP_NOW_ETH_ALEN) mp_raise_ValueError("addr invalid");
+    ESPNOW_EXCEPTIONS(esp_now_del_peer(addr_buf));
+    return mp_const_none;
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_2(espnow_del_peer_obj, espnow_del_peer);
+
+// this workaround enables ESP32 to send from whatever IF that is
+// active
+// if_id == wifi_mode - 1
+#define IS_IF_AVAILABLE(mode, if_id) ({ (mode & (if_id+1)) != 0; })
```

Comment:
I don't really see the purpose of these 2 macros, I think it would just enough to check whether the WLAN is initialized or not. Does it matter whether the mode is STA or AP ?